### PR TITLE
Add openstack charm func test runner

### PR DESCRIPTION
**Pull request by dosaboy**
Also includes tool to identify func test targets from the charm.

### CODE REVIEW

File: openstack/tools/charmed_openstack_functest_runner.sh [hunk 1]
```diff
@@ -0,0 +1,64 @@
+#!/bin/bash -ux
+source ~/novarc
+CIDR=`openstack subnet show subnet_${OS_USERNAME}-psd-extra -c cidr -f value`
+FIP_MAX=$(ipcalc $CIDR| awk '$1=="HostMax:" {print $2}')
+FIP_MIN=$(ipcalc $CIDR| awk '$1=="HostMin:" {print $2}')
+FIP_MIN_ABC=${FIP_MIN%.*}
+FIP_MIN_D=${FIP_MIN##*.}
+FIP_MIN=${FIP_MIN_ABC}.$(($FIP_MIN_D + 64))
```
Comment (brianphaley) [inline openstack/tools/charmed_openstack_functest_runner.sh:8]: This turns out to be gateway_ip + 64 since it's always .1 by default, and as long as this is a /25 or larger it's fine, smaller and it will be greater than the max. It's only until they get the gate fixed anyways...

File: openstack/tools/charmed_openstack_functest_runner.sh [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+export TEST_GATEWAY=$(openstack subnet show subnet_${OS_USERNAME}-psd-extra -c gateway_ip -f value)
+export TEST_NAME_SERVER=91.189.91.131
+export TEST_CIDR_PRIV=192.168.21.0/24
+export TEST_SWIFT_IP=10.140.56.22
+export TEST_MODEL_SETTINGS="image-stream=released;default-series=jammy;test-mode=true;transmit-vendor-metrics=false"
+export TEST_CONSTRAINTS_FILE=https://raw.githubusercontent.com/openstack-charmers/zaza/master/constraints-juju34.txt
+# We need to set TEST_JUJU3 as well as the constraints file
+# Ref: https://github.com/openstack-charmers/zaza/blob/e96ab098f00951079fccb34bc38d4ae6ebb38606/setup.py#L47
+export TEST_JUJU3=1
+export NET_ID=$(openstack network show net_${OS_USERNAME}-psd -f value -c id)
```
Comment (brianphaley) [inline openstack/tools/charmed_openstack_functest_runner.sh:23]: Just making sure this wasn't supposed to be -psd-extra network

**Comment by xtrusia** [comment]
While Im testing mysql-router charm zaza test, I realized that I need below env variable

export TEST_ZAZA_BUG_LP1987332=1

related code is here

https://github.com/openstack-charmers/zaza/blob/bef1bdbac7ee15ae742466641c65bcab363c093e/zaza/utilities/juju.py#L511

if I didn't set it, there was python array index error.

just FYI

**Comment by xtrusia** [comment]
Is there way to use PR test ? like below in commit msg

func-test-pr: https://github.com/openstack-charmers/zaza-openstack-tests/pull/1252

**Comment by dosaboy** [comment]
> While Im testing mysql-router charm zaza test, I realized that I need below env variable
> 
> export TEST_ZAZA_BUG_LP1987332=1
> 
> related code is here
> 
> https://github.com/openstack-charmers/zaza/blob/bef1bdbac7ee15ae742466641c65bcab363c093e/zaza/utilities/juju.py#L511
> 
> if I didn't set it, there was python array index error.
> 
> just FYI

I beleive this is not needed with juju3 but can add it anyway for those wanted to stick with j2

**Comment by dosaboy** [comment]
> Is there way to use PR test ? like below in commit msg
> 
> func-test-pr: [openstack-charmers/zaza-openstack-tests#1252](https://github.com/openstack-charmers/zaza-openstack-tests/pull/1252)

I have a need for this myself so would like to find a way but do not yet know how

File: openstack/tools/charmed_openstack_functest_runner.sh (hunk 2 above)
Comment (dosaboy) [reply openstack/tools/charmed_openstack_functest_runner.sh:40]: ah well spotted, this is the "external" network used for fips so for us thats \*-psd-extra

**Comment by dosaboy** [comment]
With the current version I can now do a successful build and deploy but hit the following error once the tests start:
```
2024-08-30 11:29:25 [ERROR] Model default_alias (zaza-e680c75cf4a9)
Traceback (most recent call last):
  File "/home/ubuntu/jammy/octavia/src/.tox/func-target/bin/functest-run-suite", line 8, in <module>
    sys.exit(main())
  File "/home/ubuntu/jammy/octavia/src/.tox/func-target/lib/python3.10/site-packages/zaza/charm_lifecycle/func_test_runner.py", line 405, in main
    func_test_runner(
  File "/home/ubuntu/jammy/octavia/src/.tox/func-target/lib/python3.10/site-packages/zaza/charm_lifecycle/func_test_runner.py", line 313, in func_test_runner
    run_env_deployment(env_deployment, keep_model=preserve_model,
  File "/home/ubuntu/jammy/octavia/src/.tox/func-target/lib/python3.10/site-packages/zaza/charm_lifecycle/func_test_runner.py", line 174, in run_env_deployment
    configure.configure(
  File "/home/ubuntu/jammy/octavia/src/.tox/func-target/lib/python3.10/site-packages/zaza/charm_lifecycle/configure.py", line 58, in configure
    run_configure_list(functions)
  File "/usr/lib/python3.10/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/home/ubuntu/jammy/octavia/src/.tox/func-target/lib/python3.10/site-packages/zaza/charm_lifecycle/configure.py", line 44, in run_configure_list
    utils.get_class(func)()
  File "/home/ubuntu/jammy/octavia/src/.tox/func-target/lib/python3.10/site-packages/zaza/openstack/charm_tests/neutron/setup.py", line 157, in basic_overcloud_network
    undercloud_and_charm_setup(limit_gws=limit_gws)
  File "/home/ubuntu/jammy/octavia/src/.tox/func-target/lib/python3.10/site-packages/zaza/openstack/charm_tests/neutron/setup.py", line 107, in undercloud_and_charm_setup
    network.setup_gateway_ext_port(network_config,
  File "/home/ubuntu/jammy/octavia/src/.tox/func-target/lib/python3.10/site-packages/zaza/openstack/configure/network.py", line 315, in setup_gateway_ext_port
    openstack_utils.configure_gateway_ext_port(
  File "/home/ubuntu/jammy/octavia/src/.tox/func-target/lib/python3.10/site-packages/zaza/openstack/utilities/openstack.py", line 1125, in configure_gateway_ext_port
    net_id = get_admin_net(neutronclient)['id']
TypeError: 'NoneType' object is not subscriptable
```

**Comment by dosaboy** [comment]
I have hopefully fixed the above issue which might have been caused by the fact that had forgotten to export CIDR_EXT but I also notice that both the zaza and zaza-openstack-tests repos have duplicate code for e.g. https://github.com/openstack-charmers/zaza/blob/master/zaza/utilities/generic.py#L92 https://github.com/openstack-charmers/zaza-openstack-tests/blob/master/zaza/openstack/utilities/generic.py#L153 which are slightly different in which env vars they look for and it is not clear which is actually used when a test is run so I am now setting both kinds.

**Comment by dosaboy** [comment]
works now but new issue is that nova-compute vms use a flavor too small to be able to boot any vms

**Comment by dosaboy** [comment]
> works now but new issue is that nova-compute vms use a flavor too small to be able to boot any vms

```
2024-08-30 13:40:04.351 230827 INFO nova.scheduler.manager [None req-f758d68f-0f19-461c-9068-cb4e96e272e7 184d6a2c973043d8b5ac655217b4c3cf ac3a5e14642c4a9dad8fcde1b0c3530e - - c8b548e494dd4827aa0418706da3f6a7 c8b548e494dd4827aa0418706da3f6a7] Got no allocation candidates from the Placement API. This could be due to insufficient resources or a temporary occurrence as compute nodes start up.
```

**Comment by dosaboy** [comment]
Putting this up for final review since I am now able to successfully run tests.

**Comment by rodrigogansobarbieri** [comment]
I tried running the script but I got the same error:

Juju.errors.JujuConnectionError: juju server-version 3.5.3 not supported

I changed my juju snap version to 2.9 and pointed the credentials to my juju 2.9 controller and it worked. It is not running the tests. I guess whether it supports juju 3 or not varies from charm to charm. I am running nova-compute charm tests.

One thing I noticed is that the timeout is set to 3600 instead of 7200 as Trent suggested in his notes: export TEST_DEPLOY_TIMEOUT=7200

Also one thing I don't like is that there is currently to option to specify the bundle we want to run. For example, for my nova-compute tests I am running for the 2023.2 branch, so it started running jammy-antelope, and then later it will run jammy-bobcat. Assuming jammy-bobcat fails for some reason and I have to run it again. If I have already collected a successful output from jammy-antelope I would not want to let it run again, but instead specify that I want only to run jammy-bobcat.

**Review by rodrigogansobarbieri — CHANGES_REQUESTED** [review]
-1 for lacking an option to specify the bundle to run (example: run just jammy-bobcat instead of the whole suite that would include jammy-antelope, jammy-bobcat and so on).

**Comment by dosaboy** [comment]
> -1 for lacking an option to specify the bundle to run (example: run just jammy-bobcat instead of the whole suite that would include jammy-antelope, jammy-bobcat and so on).

done.

File: openstack/tools/charmed_openstack_functest_runner.sh [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+#
+# Usage: clone/fetch charm to test and run from within charm root dir.
+#
+FUNC_TEST_PR=
+FUNC_TEST_TARGET=
+SKIP_BUILD=false
+
+usage () { echo "USAGE: --func-test-target | --func-test-pr | --skip-build"; }
+
+while (($# > 0))
```
Comment (lathiat) [inline openstack/tools/charmed_openstack_functest_runner.sh:14]: If you call the script without any arguments and you're not inside a charm directory it eventually fails. It should check for the current directory being a charm somehow.

File: openstack/tools/charmed_openstack_functest_runner.sh [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+# Run Charmed Openstack CI tests manually in a similar way to how they are run
+# by OpenStack CI (OSCI).
+#
+# Usage: clone/fetch charm to test and run from within charm root dir.
+#
+FUNC_TEST_PR=
+FUNC_TEST_TARGET=
+SKIP_BUILD=false
+
+usage () { echo "USAGE: --func-test-target | --func-test-pr | --skip-build"; }
```
Comment (lathiat) [inline openstack/tools/charmed_openstack_functest_runner.sh:12]: Ideally this would make it more clear how to use the tool, e.g. that you have to call it from within a charm directory, etc.

File: openstack/tools/charmed_openstack_functest_runner.sh [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+fi
+
+for target in ${func_targets[@]}; do
+    [[ -d src ]] && pushd src || true
+    tox -re func-target -- $target
+
+    read -p "Destroy model and run next test? [ENTER]"
+    # cleanup before next run
+    model=`juju list-models| egrep -o "^zaza-\S+"|tr -d '*'`
+    juju destroy-model --no-prompt $model --force --no-wait --destroy-storage
```
Comment (lathiat) [inline openstack/tools/charmed_openstack_functest_runner.sh:119]: This will fail if you have multiple matching models. Ideally it would also print the name of all the models it will destroy before it does, in case you had created another model with a name starting with zaza.

You can also just tell zaza to destroy the model itself automatically by passing an argument, perhaps we should use that by default and then have an option to switch to manual cleanup behaviour for trying to debug the tests?

**Review by lathiat — CHANGES_REQUESTED** [review]
These are nits, and not really related to the tool's core functionality. But thought they are worth adding/fixing.

**Comment by lathiat** [comment]
Tested working OK for me with glance-simplestreams-sync yoga

File: openstack/tools/charmed_openstack_functest_runner.sh (hunk 5 above)
Comment (dosaboy) [reply openstack/tools/charmed_openstack_functest_runner.sh:119]: do you happen to know what the argument is?

Comment (dosaboy) [reply openstack/tools/charmed_openstack_functest_runner.sh:157]: we can fix this as a future patch

File: openstack/tools/charmed_openstack_functest_runner.sh (hunk 4 above)
Comment (dosaboy) [reply openstack/tools/charmed_openstack_functest_runner.sh]: i have put in a much nicer help message

File: openstack/tools/charmed_openstack_functest_runner.sh (hunk 3 above)
Comment (dosaboy) [reply openstack/tools/charmed_openstack_functest_runner.sh:46]: i will fix this as a followup patch but have put info in the help message about how to run the tool
